### PR TITLE
feat:[CDS-59065]: Service dashboard revamp phase 2

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes

These changes corresponds to service dashboard revamp phase 2.

https://harness.atlassian.net/wiki/spaces/CDNG/pages/21274918929/Service+Details+Page+Enhancements

Test Plan: https://docs.google.com/spreadsheets/d/1stT-GmpsbCmb6qUFL8Nv-JnV4iRfYN7BDKY-5u285pE/edit#gid=538678939

Mocks : https://www.figma.com/file/ZMUBknquCgGI12eEl6394W/Service-%26-Environment-Overviews?node-id=7466-301953&t=iRAYixh8FaiC0FJq-0

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46729)

### CODE REVIEW

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/overview/dto/ServicePipelineWithRevertInfo.java` line 22 at r10](https://reviewable.io/reviews/harness/harness-core/46729#-NTmjmSF2IHaMARCC9JI:-NTmjmSF2IHaMARCC9JJ:bup0upw) ([raw file](https://github.com/harness/harness-core/blob/39a6712b03ad91e43aa897a5c8d5dbb9b3588bf0/120-ng-manager/src/main/java/io/harness/ng/overview/dto/ServicePipelineWithRevertInfo.java#L22)):*
> ```Java
> @OwnedBy(HarnessTeam.DX)
> @Value
> @Getter
> ```

Getter is auto generated by Value

**Review by yogesh-chauhan — DISMISSED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 1]
```diff
@@ -1477,6 +1556,29 @@ public List<String> getPipelineExecutionIdFromServiceInfraInfo(String query) {
     return ids.stream().collect(Collectors.toList());
   }
 
+  public List<String> getPipelineExecutionsWhereRollbackOccurred(String query) {
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:1559]: there is existing method 'getPipelineExecutionIdFromServiceInfraInfo' that can be used

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 2]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+      String orgIdentifier, String projectIdentifier, String serviceId, String environmentId, String envGrpId) {
     boolean isGitOps = isGitopsEnabled(accountIdentifier, orgIdentifier, projectIdentifier, serviceId);
     List<ActiveServiceInstanceInfoWithEnvType> activeServiceInstanceInfoList =
         instanceDashboardService.getActiveServiceInstanceInfoWithEnvType(
             accountIdentifier, orgIdentifier, projectIdentifier, environmentId, serviceId, null, isGitOps, false);
     DashboardServiceHelper.sortActiveServiceInstanceInfoWithEnvTypeList(activeServiceInstanceInfoList);
-    return DashboardServiceHelper.getInstanceGroupedByEnvironmentListHelper(activeServiceInstanceInfoList, isGitOps);
+
+    List<String> envIds = new ArrayList<>();
+    activeServiceInstanceInfoList.forEach(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2807]: Earlier we used to show the state from instanceNG collection in tabular view, why we are comparing data with current state and discarding

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 3]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+        activeServiceInstanceInfo -> envIds.add(activeServiceInstanceInfo.getEnvIdentifier()));
+
+    Criteria criteria = environmentGroupService.formCriteria(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2810]: why we are not fetching specific envGroup details?

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 4]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+        accountIdentifier, orgIdentifier, projectIdentifier, false, "", "", null, false);
+    Page<EnvironmentGroupEntity> environmentGroupEntitiesPage =
+        environmentGroupService.list(criteria, Pageable.unpaged(), projectIdentifier, orgIdentifier, accountIdentifier);
+
+    List<String> scopeSpecificEnvironmentRefs = environmentService.fetchesNonDeletedEnvIdentifiersFromList(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2815]: environments can have different scopes, use method which fetches based on refs

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 5]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+
+    activeServiceInstanceInfoList = filterNonDeletedEnvs(activeServiceInstanceInfoList, environments);
+
+    return DashboardServiceHelper.getInstanceGroupedByEnvironmentListHelper(
+        envGrpId, activeServiceInstanceInfoList, isGitOps, environmentGroupEntitiesPage);
+  }
+
+  private List<ActiveServiceInstanceInfoWithEnvType> filterNonDeletedEnvs(
+      List<ActiveServiceInstanceInfoWithEnvType> activeServiceInstanceInfoList, List<Environment> environments) {
+    List<String> envIds = new ArrayList<>();
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2828]: envRef should be basis of filtering

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 6]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
-    return DashboardServiceHelper.getInstanceGroupedByArtifactListHelper(activeServiceInstanceInfoList, isGitOps);
+
+    List<String> envIds = new ArrayList<>();
+    activeServiceInstanceInfoList.forEach(
+        activeServiceInstanceInfo -> envIds.add(activeServiceInstanceInfo.getEnvIdentifier()));
+
+    Criteria criteria = environmentGroupService.formCriteria(
+        accountIdentifier, orgIdentifier, projectIdentifier, false, "", "", null, false);
+    Page<EnvironmentGroupEntity> environmentGroupEntitiesPage =
+        environmentGroupService.list(criteria, Pageable.unpaged(), projectIdentifier, orgIdentifier, accountIdentifier);
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2865]: Comments similar to InstanceGroupedByEnvironment method

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 7]
```diff
@@ -3005,32 +3236,199 @@ public List<InstanceGroupedByServiceList.InstanceGroupedByPipelineExecution> gro
...
+    if (environmentGroupEntitiesPage != null) {
+      environmentGroupEntities = environmentGroupEntitiesPage.getContent();
+      for (EnvironmentGroupEntity environmentGroupEntity : environmentGroupEntities) {
+        if (EmptyPredicate.isNotEmpty(environmentGroupEntity.getEnvIdentifiers())) {
+          envIds.addAll(environmentGroupEntity.getEnvIdentifiers());
+        }
+      }
+    }
+
+    List<String> scopeSpecificEnvironmentRefs = environmentService.fetchesNonDeletedEnvIdentifiersFromList(
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3266]: comments similar to InstanceGroupedByEnvironment method

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 8]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+        activeServiceInstanceInfo -> envIds.add(activeServiceInstanceInfo.getEnvIdentifier()));
+
+    Criteria criteria = environmentGroupService.formCriteria(
+        accountIdentifier, orgIdentifier, projectIdentifier, false, "", "", null, false);
+    Page<EnvironmentGroupEntity> environmentGroupEntitiesPage =
+        environmentGroupService.list(criteria, Pageable.unpaged(), projectIdentifier, orgIdentifier, accountIdentifier);
+
+    List<String> scopeSpecificEnvironmentRefs = environmentService.fetchesNonDeletedEnvIdentifiersFromList(
+        accountIdentifier, orgIdentifier, projectIdentifier, new ArrayList<>(envIds));
+    List<Environment> environments = environmentService.fetchesNonDeletedEnvironmentFromListOfRefs(
```
[the remark this answers is not in the document]
Comment (piyush-bhuwalka) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2817]: We'll check this

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 5 above)
Comment (piyush-bhuwalka) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2828]: We'll check this

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 9]
```diff
@@ -3041,20 +3403,52 @@ public ArtifactInstanceDetails getArtifactInstanceDetails(
...
+    Page<EnvironmentGroupEntity> environmentGroupEntitiesPage =
+        environmentGroupService.list(criteria, Pageable.unpaged(), projectIdentifier, orgIdentifier, accountIdentifier);
+
+    List<EnvironmentGroupEntity> environmentGroupEntities = null;
+
+    if (environmentGroupEntitiesPage != null) {
+      environmentGroupEntities = environmentGroupEntitiesPage.getContent();
+      for (EnvironmentGroupEntity environmentGroupEntity : environmentGroupEntities) {
+        if (EmptyPredicate.isNotEmpty(environmentGroupEntity.getEnvIdentifiers())) {
+          envIds.addAll(environmentGroupEntity.getEnvIdentifiers());
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3441]: why we are adding these envIds to list?

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 10]
```diff
@@ -3005,32 +3213,186 @@ public List<InstanceGroupedByServiceList.InstanceGroupedByPipelineExecution> gro
...
+    Map<String, ServicePipelineWithRevertInfo> pipelineExecutionDetailsMap = getPipelineExecutionDetailsWithRevertInfo(
+        artifactDeploymentDetailsMap.values()
+            .stream()
+            .filter(artifactDeploymentDetail
+                -> EmptyPredicate.isNotEmpty(artifactDeploymentDetail.getLastPipelineExecutionId()))
+            .map(artifactDeploymentDetail -> artifactDeploymentDetail.getLastPipelineExecutionId())
+            .collect(Collectors.toList()));
+    List<String> pipelineExecutionIdsWhereRollbackOccurred = getPipelineExecutionsWhereRollbackOccurred(
+        pipelineExecutionDetailsMap.values()
+            .stream()
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3252]: why we are fetching pipeline execution id from this map and not similar to above? also pipeline identifier list is provided where as in query filtering is done on pipeline_execution_summary_cd_id, is this tested?

Comment (piyush-bhuwalka) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:3253]: Done

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 2 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2784]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 11]
```diff
@@ -2624,24 +2796,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
             accountIdentifier, orgIdentifier, projectIdentifier, environmentId, serviceId, null, isGitOps, false);
     DashboardServiceHelper.sortActiveServiceInstanceInfoWithEnvTypeList(activeServiceInstanceInfoList);
-    return DashboardServiceHelper.getInstanceGroupedByEnvironmentListHelper(activeServiceInstanceInfoList, isGitOps);
+
+    List<String> envIds = new ArrayList<>();
+    activeServiceInstanceInfoList.forEach(
+        activeServiceInstanceInfo -> envIds.add(activeServiceInstanceInfo.getEnvIdentifier()));
+
+    Criteria criteria = environmentGroupService.formCriteria(
+        accountIdentifier, orgIdentifier, projectIdentifier, false, "", "", null, false);
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2788]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 4 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 5 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2806]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 12]
```diff
@@ -2624,24 +2773,83 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+      List<ActiveServiceInstanceInfoWithEnvType> activeServiceInstanceInfoList, List<Environment> environments) {
+    List<String> envIds = new ArrayList<>();
+    HashMap<String, Environment> envRefEnvMap = new HashMap<>();
+    environments.forEach(environment -> {
+      envRefEnvMap.put(IdentifierRefHelper
+                           .getIdentifierRefWithScope(environment.getAccountId(), environment.getOrgIdentifier(),
+                               environment.getProjectIdentifier(), environment.getIdentifier())
+                           .buildScopedIdentifier(),
+          environment);
+      envIds.add(environment.getIdentifier());
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 9 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 3 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2787]: resolved

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java (hunk 6 above)
Comment (sourabh-awashti) [reply 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2845]: resolved

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 17 files at r1, 8 of 20 files at r2, 2 of 4 files at r4, 1 of 4 files at r5, 8 of 9 files at r7, 6 of 9 files at r8, 2 of 3 files at r9, 2 of 5 files at r11, 1 of 2 files at r14, 2 of 2 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46729)* status: 32 file(s) to review. 1 change requests. 12. unresolved discussions. 1 of 2 approvals obtained.

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 13]
```diff
@@ -2624,24 +2773,84 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
-    return DashboardServiceHelper.getInstanceGroupedByEnvironmentListHelper(activeServiceInstanceInfoList, isGitOps);
+
+    List<String> envIds = new ArrayList<>();
+    activeServiceInstanceInfoList.forEach(
+        activeServiceInstanceInfo -> envIds.add(activeServiceInstanceInfo.getEnvIdentifier()));
+
+    Criteria criteria = environmentGroupService.formCriteria(
+        accountIdentifier, orgIdentifier, projectIdentifier, false, "", "", null, false);
+    Page<EnvironmentGroupEntity> environmentGroupEntitiesPage =
+        environmentGroupService.list(criteria, Pageable.unpaged(), projectIdentifier, orgIdentifier, accountIdentifier);
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2790]: Environments belonging to envGroups are not fetched using method that you created

File: 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java [hunk 14]
```diff
@@ -2624,24 +2773,84 @@ public EnvBuildIdAndInstanceCountInfoList getEnvBuildInstanceCountByServiceId(
...
+
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/overview/service/CDOverviewDashboardServiceImpl.java:2846]: comment same as above

File: 127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceSequenceRepositoryCustomImpl.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  public Update getUpdateOperations(Update update, ServiceSequence serviceSequence) {
+    update.set(ServiceSequenceKeys.accountId, serviceSequence.getAccountId());
+    update.set(ServiceSequenceKeys.orgIdentifier, serviceSequence.getOrgIdentifier());
+    update.set(ServiceSequenceKeys.projectIdentifier, serviceSequence.getProjectIdentifier());
+    update.set(ServiceSequenceKeys.serviceIdentifier, serviceSequence.getServiceIdentifier());
+    update.setOnInsert(ServiceSequenceKeys.createdAt, System.currentTimeMillis());
+    update.set(ServiceSequenceKeys.lastModifiedAt, System.currentTimeMillis());
+    update.set(ServiceSequenceKeys.serviceIdentifier, serviceSequence.getServiceIdentifier());
```
Comment (piyush-bhuwalka) [inline 127-cd-nextgen-entities/src/main/java/io/harness/repositories/service/custom/ServiceSequenceRepositoryCustomImpl.java:68]: remove this, looks duplicate

**Pull request merged**